Mathlib/CategoryTheory/Limits/Types/ColimitTypeFiltered.lean

```diff
@@ -0,0 +1,118 @@
+Given `c : CoconeTypes F`, we deduce a lemma
+`Functor.CoconeTypes.injective_descColimitType_iff_of_isFiltered` which gives
+a concrete condition under which the map
+`F.descColimitType c : F.ColimitType → c.pt` is injective, which is an
+important step when proving `c.IsColimit`.
+
+-/
+
+universe w₁ w₀ v u
+
+namespace CategoryTheory
+
+open IsFiltered
+
+variable {J : Type u} [Category.{v} J] [IsFiltered J]
+
+namespace Functor
+
+variable (F : J ⥤ Type w₀)
+
+lemma eqvGen_colimitTypeRel_iff_of_isFiltered
+    (x y : (j : J) × (F.obj j)) :
+    Relation.EqvGen F.ColimitTypeRel x y ↔
+      ∃ (k : J) (f : x.1 ⟶ k) (g : y.1 ⟶ k), F.map f x.2 = F.map g y.2 := by
+  constructor
+  · intro h
+    induction h with
+    | rel x y h =>
+      obtain ⟨f, h⟩ := h
+      exact ⟨y.1, f, 𝟙 _, by simpa using h.symm⟩
+    | refl x => exact ⟨x.1, 𝟙 _, 𝟙 _, rfl⟩
+    | symm _ _ _ h =>
+      obtain ⟨k, f, g, h⟩ := h
+      exact ⟨k, g, f, h.symm⟩
+    | trans x y z _ _ h h' =>
+      obtain ⟨k, f, g, h⟩ := h
+      obtain ⟨k', f', g', h'⟩ := h'
+      obtain ⟨l, a, b, h''⟩ := span g f'
+      refine ⟨l, f ≫ a, g' ≫ b, ?_⟩
+      rw [FunctorToTypes.map_comp_apply, FunctorToTypes.map_comp_apply _ g', h, ← h',
+        ← FunctorToTypes.map_comp_apply, ← FunctorToTypes.map_comp_apply, h'']
+  · rintro ⟨k, f, f', h⟩
+    apply Relation.EqvGen.trans (y := ⟨k, F.map f' y.2⟩)
+    · exact .rel _ _ ⟨f, by rw [← h]⟩
+    · exact .symm _ _ (.rel _ _ ⟨f', rfl⟩)
+
+lemma ιColimitType_eq_iff_of_isFiltered {j j' : J} (x : F.obj j) (y : F.obj j') :
+    F.ιColimitType j x = F.ιColimitType j' y ↔
+      ∃ (k : J) (f : j ⟶ k) (f' : j' ⟶ k), F.map f x = F.map f' y := by
+  rw [ιColimitType_eq_iff, eqvGen_colimitTypeRel_iff_of_isFiltered]
+
+lemma ιColimitType_eq_iff_of_isFiltered' {j : J} (x y : F.obj j) :
+    F.ιColimitType j x = F.ιColimitType j y ↔
+      ∃ (k : J) (f : j ⟶ k), F.map f x = F.map f y := by
+  rw [ιColimitType_eq_iff_of_isFiltered]
+  constructor
+  · rintro ⟨k, f, f', h⟩
+    refine ⟨coeq f f', f ≫ coeqHom f f', ?_⟩
+    nth_rw 2 [coeq_condition]
+    simp [h]
+  · rintro ⟨k, f, h⟩
+    exact ⟨k, f, f, h⟩
+
+namespace CoconeTypes
+
+variable {F} (c : CoconeTypes.{w₁} F)
+
+lemma injective_descColimitType_iff_of_isFiltered :
+    Function.Injective (F.descColimitType c) ↔
+      ∀ (j j' : J) (x : F.obj j) (x' : F.obj j'), c.ι j x = c.ι j' x' →
+        ∃ (k : J) (f : j ⟶ k) (f' : j' ⟶ k), F.map f x = F.map f' x' := by
+  constructor
+  · intro h j j' x x' eq
+    have : F.ιColimitType j x = F.ιColimitType j' x' := h eq
+    rwa [ιColimitType_eq_iff_of_isFiltered] at this
+  · intro h x x' eq
+    obtain ⟨i, x, rfl⟩ := F.ιColimitType_jointly_surjective x
+    obtain ⟨i', x', rfl⟩ := F.ιColimitType_jointly_surjective x'
+    simp only [descColimitType_ιColimitType_apply] at eq
+    obtain ⟨k, f, f', eq⟩ := h _ _ _ _ eq
+    rw [← F.ιColimitType_map f x, eq, F.ιColimitType_map]
+
+lemma injective_descColimitType_iff_of_isFiltered' :
```

Comment:
```suggestion
lemma descColimitType_injective_iff_of_isFiltered' :
```
(do we have a better way to distinguish these than adding a prime? Also, please add a documentation comment to both of them to explain the difference)